nalgebra-sparse/src/io/matrix_market.rs

```diff
@@ -0,0 +1,861 @@
+                    MMErrorKind::TypeUnmatched,
+                    format!("Pattern type has no negative"),
+                ))
+            }
+        }
+    };
+}
+
+mm_int_impl!(i8);
+mm_int_impl!(i16);
+mm_int_impl!(i32);
+mm_int_impl!(i64);
+mm_int_impl!(i128);
+
+mm_float_impl!(f32);
+mm_float_impl!(f64);
+
+mm_complex_impl!(f32);
+mm_complex_impl!(f64);
+
+mm_pattern_impl!(());
+
+#[derive(Parser)]
+#[grammar = "io/matrix_market.pest"]
+struct MMParser;
+
+/// Parsing a pest structure to a Typecode of the matrix.
+fn parsing_header(inner: &mut Pairs<'_, Rule>, header_type: &mut Typecode) -> Result<(), MMError> {
+    // once the data can be parsed, all unwrap() in this function here will not panic.
+    header_type.sp = inner
+        .next()
+        .unwrap()
+        .as_str()
+        .to_ascii_lowercase()
+        .parse::<Sparsity>()
+        .unwrap();
+    header_type.dt = inner
+        .next()
+        .unwrap()
+        .as_str()
+        .to_ascii_lowercase()
+        .parse::<DataType>()
+        .unwrap();
+    header_type.ss = inner
+        .next()
+        .unwrap()
+        .as_str()
+        .to_ascii_lowercase()
+        .parse::<StorageScheme>()
+        .unwrap();
+    typecode_precheck(&header_type)
+}
+
+/// Parsing a pest structure to 3 int, which are number of rols, cols and non-zeros.
+fn parsing_shape(inner: &mut Pairs<'_, Rule>, shape: &mut (usize, usize, usize)) {
+    // once the file can be parsed, all unwrap() in this function here will not panic.
+    shape.0 = inner.next().unwrap().as_str().parse::<usize>().unwrap();
+    shape.1 = inner.next().unwrap().as_str().parse::<usize>().unwrap();
+    shape.2 = inner.next().unwrap().as_str().parse::<usize>().unwrap();
+}
+
+#[inline]
+/// Do a precheck of remaing data. Because in matrix_market.pest file, either 0, or 1 or 2 float number(s) as data is valid, however, for example, for Real matrix, only 1 float is valid. And 2 float for Complex matrix.  
+fn check_value_length(inner: &Pairs<'_, Rule>, l: usize) -> Result<(), MMError> {
+    let copy = inner.clone();
+    let c = copy.count();
+    if l != c {
+        return Err(MMError::from_kind_and_message(
+            MMErrorKind::ParsingError,
+            format!("{} data required, but {} provided.", l, c),
+        ));
+    }
+    Ok(())
+}
+
+#[inline]
+// do a quick check it the entry is in the lower triangle part of the matrix
+fn check_lower_triangle(r: usize, c: usize) -> Result<(), MMError> {
+    if c > r {
+        return Err(MMError::from_kind_and_message(
+            MMErrorKind::LowerTriangleError,
+            format!(
+                "Entry: row {} col {} should be put into lower triangle",
+                r, c
+            ),
+        ));
+    }
+    Ok(())
+}
+
+/// Parses a Matrix Market file at the given path, and returns the corresponding sparse matrix as CooMatrix format.
+pub fn load_coo_from_mm_file<T, P: AsRef<Path>>(path: P) -> Result<CooMatrix<T>, MMError>
```

Comment:
Maybe `load_coo_from_matrix_market_file`?

Reply:
Maybe also add an `# Errors` section containing something like:

> # Errors
> See [`MatrixMarketErrorKind`] for a list of possible error conditions.

And the same to `_from_mm_str`?